### PR TITLE
feat!: move closer to test double definitions

### PR DESCRIPTION
**Pull request by crookse**
## Overview of commits

- feat: add Fake()
- feat: add Dummy()
- feat: new mock methods (expects(), method(), verifyExpectations())
- feat: pre-programmed method implementation for faking/mocking methods
- feat: add interfaces so that they can be returned for improved type hinting
- refactor: mocks to use mixin
- refactor!: stubs (no need to stub a class to stub its members)
- refactor: types are now stricter and provide better type hinting
- test: test fakes, dummies, mocks, stubs and move tests files into unit/mod

## Updated README

See the new README [here](https://github.com/drashland/rhum/blob/feat/new-mock-methods/README.md).

## Refactor and add features to be closer to test double definitions

Test double definitions can be found here: https://martinfowler.com/bliki/TestDouble.html

Currently, we only have:
- Mocks
- Dummies
- Fakes
- Stubs (breaking changes made in this PR)

Still need to add in spies.

## Stub()

Before, we created a stub via `Stub(someObject)` and then we could call `someObject.stub(...)`. This didn't make sense per definition. The definition is:

> Stubs provide canned answers to calls made during the test, usually not responding at all to anything outside what's programmed in for the test.

We shouldn't have to call stub on an object to make it stub-able. Instead we should call `Stub(theObjectReceivingTheStub, theDataMemberToStub, optionalValueTheStubShouldReturn)`. API looks like the below now:

```typescript
class Server {
  public greeting = "hello";

  public methodThatLogs() {
    console.log("Server running.");
  }
}

const server = new Server();

Stub(server, "greeting", "you got changed");
assertEquals(server.greeting, "you got changed");

Stub(server, "greeting");
assertEquals(server.greeting, null);

// `is_stubbed` should be added when stubbing an object
assertEquals("is_stubbed" in server, true);

```

## Use mixin for mock

Introduces using a [mixin](https://www.typescriptlang.org/docs/handbook/mixins.html) to create a mock so that calling private methods and returning `this` also works in a mock. For example, the old mock implementation wouldn't work when `someComplexMethod()` below would return `this` and call `this.#setSomethingOne()` and `this.#setSomethingTwo()` because the context of `this` was the mock object and not the original object. It needs to be an instance of the original object. That's where the mixin comes in. Basically, mocks are now extensions of the original so that `mock instanceof original === true`. This approach also fixes the getter/setter issue from https://github.com/drashland/rhum/pull/148/files.

```typescript
class TestObjectFourBuilder {
  #something_one?: string;
  #something_two?: string;
  ...
  ...
  ...
  someComplexMethod(): this {
    this.#setSomethingOne();
    this.#setSomethingTwo();
    return this;
  }
  ...
  ...
  ...
}
```

## Add mock.method(...).willReturn(...)

```typescript
class TestObjectThree {
  public test(): string { return "World"; }
}

const mock = new MockBuilder(TestObjectThree).create();
assertEquals(mock.is_mock, true);

// Original returns "World"
assertEquals(mock.test(), "World");

// Don't fully pre-program the method. This should cause an error during
// assertions.
mock
  .method("test")
  .willReturn({
    name: "something"
  });

assertEquals(mock.test(), {name: "something"});
assertEquals(mock.calls.test, 2);
assertEquals(mock.calls.hello, 2);
```

## Add mock.method(...).willThrow(...)

```typescript
class TestObjectThree {
  public test(): string { return "World"; }
}

const mock = new MockBuilder(TestObjectThree).create();
assertEquals(mock.is_mock, true);

// Original returns "World"
assertEquals(mock.test(), "World");

// Make the original method throw RandomError
mock
  .method("test")
  .willThrow(new RandomError("Random error message."))

assertThrows(
  () => mock.test(),
  RandomError,
  "Random error message."
);
assertEquals(mock.calls.test, 2);
```

## Add mock.expects(...).toBeCalled(...)

```typescript
const mock = new MockBuilder(TestObjectThree).create();
assertEquals(mock.is_mock, true);

mock.expects("hello").toBeCalled(2);

mock.test(); // Calls .hello() under the hood twice

mock.verifyExpectations(); // Must be called to verify the .expects() calls
```

## Add Fake()

Fakes kind of act like mocks, but they do not have calls and expectations like mocks.

```typescript
// Assert that a fake can make a class take a shortcut
const fakeServiceDoingShortcut = Fake(Repository).create();
fakeServiceDoingShortcut.method("findAllUsers").willReturn("shortcut");
const resourceWithShortcut = new Resource(
  fakeServiceDoingShortcut,
);
resourceWithShortcut.getUsers();
assertEquals(fakeServiceDoingShortcut.anotha_one_called, false);
assertEquals(fakeServiceDoingShortcut.do_something_called, false);
assertEquals(fakeServiceDoingShortcut.do_something_else_called, false);

// Assert that the fake service is not yet doing a shortcut
const fakeServiceNotDoingShortcut = Fake(Repository).create();
const resourceWithoutShortcut = new Resource(
  fakeServiceNotDoingShortcut,
);
resourceWithoutShortcut.getUsers();
assertEquals(fakeServiceNotDoingShortcut.anotha_one_called, true);
assertEquals(fakeServiceNotDoingShortcut.do_something_called, true);
assertEquals(fakeServiceNotDoingShortcut.do_something_else_called, true);

```

## Add Dummy()

Dummies are empty objects of an instance of something. For example, `Dummy(Hello) instanceof Hello === true`. This makes it a little bit easier to fill in parameter lists where parameters must be instances of something.

```typescript
const mockServiceOne = Mock(ServiceOne).create();
const dummy3 = Dummy(ServiceThree);

const resource = new Resource(
  mockServiceOne,
  Dummy(ServiceTwo),
  dummy3,
);

resource.callServiceOne();
assertEquals(mockServiceOne.calls.methodServiceOne, 1);
```

Dummies can also be created without having to specify constructor arguments.

```typescript
class Hello {
  constructor(
    arg1: SomethingOne,
    arg2: SomethingTwo,
    arg3: SomethingThree,
  ) {
    ...
    ...
    ...
  }
}

const dummy = Dummy(Hello); // works without throwing error
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,47 +1,15 @@
...
-    <img src="https://img.shields.io/badge/chat-on%20discord-blue">
-  </a>
-  <a href="https://twitter.com/drash_land">
-    <img src="https://img.shields.io/twitter/url?label=%40drash_land&style=social&url=https%3A%2F%2Ftwitter.com%2Fdrash_land">
-  </a>
-  <a href="https://www.youtube.com/watch?v=WhG5hLrcaVQ&list=PLlFUbR9MhiNU9VlCi97JkahXyDYcL_vUz&ab_channel=drashland">
-    <img src="https://img.shields.io/badge/Tutorials-YouTube-red">
-  </a>
-</p>
+# Drash
```
Comment (ebebbington) [inline README.md:1]: *cough*

File: extend.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// deno-fmt-ignore-file
+// deno-lint-ignore-file
+// This code was bundled using `deno bundle` and it's not recommended to edit it manually
+
+class Base {
```
Comment (ebebbington) [inline extend.js:5]: wha dis doing here ;)

File: extend.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class Base {
+  public base_prop = "base prop";
+}
```
Comment (ebebbington) [inline extend.ts:3]: same as above, test file you committed?

File: src/mock/mock_mixin.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    //////////////////////////////////////////////////////////////////////////////
+
+    /**
+     * @param original - The original object to mock.
+     * @param methodsToTrack - The original object's method to make trackable.
+     */
+    public init(original: OriginalObject, methodsToTrack: string[]) {
+      this.#original = original;
+      this.#calls = this.#constructCallsProperty(methodsToTrack);
+    }
```
Comment (ebebbington) [inline src/mock/mock_mixin.ts:67]: constructor needs to be updated with methods - public

Comment (ebebbington) [reply src/mock/mock_mixin.ts:67]: actually that is below so maybe rm that comment block and move method?

**Review by ebebbington — CHANGES_REQUESTED** [review]
Omg this is beautiful eric, usual minor things but my god the code is sexy

File: README.md (hunk 1 above)
Comment (crookse) [reply README.md:1]: HAHA dang.

File: extend.js (hunk 2 above)
Comment (crookse) [reply extend.js:5]: LOL test files

File: README.md (hunk 1 above)
Comment (crookse) [reply README.md]: doneski: [cc0ebdf](https://github.com/drashland/rhum/pull/150/commits/cc0ebdf2313311e28eb185fd605943576b5c3616)

File: extend.js (hunk 2 above)
Comment (crookse) [reply extend.js]: donezo: [92bd978](https://github.com/drashland/rhum/pull/150/commits/92bd978cb4dff8f81caf474588f116654aab3f8a)

File: extend.ts (hunk 3 above)
Comment (crookse) [reply extend.ts]: dooonnnne: [92bd978](https://github.com/drashland/rhum/pull/150/commits/92bd978cb4dff8f81caf474588f116654aab3f8a)

File: src/mock/mock_mixin.ts (hunk 4 above)
Comment (crookse) [reply src/mock/mock_mixin.ts]: moved them stuffs: [08e47ef](https://github.com/drashland/rhum/pull/150/commits/08e47efb29339ffe1b1a8044ef45b9e9b07bb361)

**Review by ebebbington — APPROVED** [review]

**Pull request merged**